crowbar_framework/app/models/backup.rb

```diff
@@ -0,0 +1,181 @@
+#
+# Copyright 2015, SUSE LINUX GmbH
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+require "find"
+
+class Backup
+  include ActiveModel::Model
+
+  attr_accessor :name, :created_at, :filename, :path
+
+  validates_presence_of :name, :created_at
+  validates :created_at, exclusion: { in: [nil, ""] }
+  validate :filename, :filename_has_format
+  validate :filename, :filename_has_characters
+
+  def initialize(name, created_at)
+    @name = name
+    @created_at = created_at
+    @filename = "#{@name}-#{@created_at}.tar.gz"
+    @path = "#{Backup.image_dir}/#{@filename}"
+  end
+
+  def save
+    valid?
+  end
+
+  def exist?
+    !Backup.where(name: name, created_at: created_at).nil?
+  end
+
+  def delete
+    File.delete(@path)
+  end
+
+  def upload(file)
+    File.open("#{Backup.image_dir}/#{file.original_filename}", "wb") do |f|
+      f.write(file.read)
+    end
+  end
+
+  def restore
+    upgrade if upgrade?
+    ret = data_valid?
+    return ret unless ret[:status] == :ok
+    Crowbar::Backup::Restore.new(self).restore
+  end
+
+  def extract
+    backup_dir = Dir.mktmpdir
+    Archive.extract(path.to_s, backup_dir)
+    Pathname.new(backup_dir)
+  end
+
+  def data
+    @data ||= extract
+  end
+
+  def data_valid?
+    validate = Crowbar::Backup::Validate.new(data)
+    validate.validate
+  end
+
+  def version
+    @version ||= begin
+      version_file = data.join("crowbar", "version")
+      File.open(version_file, &:readline).match("[0-9]*\.[0-9]*").to_s
+    end
+  end
+
+  def upgrade?
+    ENV["CROWBAR_VERSION"] > version
+  end
+
+  def upgrade
+    upgrade = Crowbar::Upgrade.new(self)
+    if upgrade.supported?
+      upgrade.upgrade
+    else
+      return { status: :not_acceptable, msg: I18n.t(".upgrade_not_supported", scope: "backup.index") }
```

Comment:
Metrics/LineLength: Line is too long. [102/100] (https://github.com/bbatsov/ruby-style-guide#80-character-limits)